examples/07_end_to_end/earthquake.ipynb

```diff
@@ -0,0 +1,416 @@
+    "        self._thread.start()\n",
+    "        self._starttime = datetime.utcnow()\n",
+    "\n",
+    "    def stop(self):\n",
+    "        print(\"FetchEventDataAdapter::stop\")\n",
+    "        if self._running:\n",
+    "            self._running = False\n",
+    "            self._thread.join()\n",
+    "\n",
+    "    def _run(self):\n",
+    "        history = []\n",
+    "        while self._running:\n",
+    "            now = datetime.utcnow()\n",
+    "            print(\"----------------------------------------------\")\n",
+    "            print(f\"{now}: refreshing earthquake feed\")\n",
+    "            print(\"----------------------------------------------\")\n",
+    "            catalog = read_events(self._url, format=\"QUAKEML\")\n",
+    "            def sorting(event):\n",
+    "                return event.origins[0].time\n",
+    "            catalog.events.sort(key=sorting)\n",
+    "            # tick whenever new events appear\n",
+    "            live = self._realtime\n",
+    "            if live:\n",
+    "                for event in catalog:\n",
+    "                    if event not in history:\n",
+    "                        history.append(event)\n",
+    "                        my_event = EventData(\n",
+    "                            time=event.origins[0].time,\n",
+    "                            longitude=event.origins[0].longitude,\n",
+    "                            latitude=event.origins[0].latitude,\n",
+    "                            magnitude=event.magnitudes[0].mag,\n",
+    "                            creation_time=event.origins[0].creation_info.creation_time.datetime,\n",
+    "                        )\n",
+    "                        # For a PushPullAdapter, push_tick takes 3 arguments: push_tick(live: bool, time: datetime, value: csp.Struct)\n",
+    "                        self.push_tick(live, my_event.time.datetime, my_event)\n",
+    "                interval = self._interval.total_seconds()\n",
+    "                time.sleep(interval)\n",
+    "            else:\n",
+    "                for event in catalog:\n",
+    "                    time.sleep(0.2)\n",
+    "                    if event not in history:\n",
+    "                        history.append(event)\n",
+    "                        my_event = EventData(\n",
+    "                            time=event.origins[0].time,\n",
+    "                            longitude=event.origins[0].longitude,\n",
+    "                            latitude=event.origins[0].latitude,\n",
+    "                            magnitude=event.magnitudes[0].mag,\n",
+    "                            creation_time=event.origins[0].creation_info.creation_time.datetime,\n",
+    "                        )\n",
+    "                        # For a PushPullAdapter, push_tick takes 3 arguments: push_tick(live: bool, time: datetime, value: csp.Struct)\n",
+    "                        self.push_tick(live, now, my_event)\n",
```

Comment:
This can probably be improved quite a lot. I am not sure if this is the best approach to define both branches.